io-sim/test/Test/STM.hs

```diff
@@ -0,0 +1,795 @@
+    Env (Map.insert name (SomeValue v) env)
+
+
+--
+-- Heaps for mutable variable
+--
+
+data SomeVar where
+     SomeVar :: Var t -> SomeVar
+
+-- | The heap is a mapping of 'Var's to their current values.
+--
+newtype Heap = Heap (Map VarId SomeValue)
+  deriving (Show, Semigroup, Monoid)
+
+-- | The STM semantics uses two heaps, the other one is called the allocations.
+type Allocs = Heap
+
+
+readVar :: Heap -> Var t -> Value t
+readVar (Heap heap) (Var n tyrep) =
+    fromMaybe (error "readVar: no such var") $ do
+    SomeValue v <- Map.lookup n heap
+    Refl        <- tyrep `eqTyRep` tyRepValue v
+    return v
+
+writeVar :: Heap -> Var t -> Value t -> Heap
+writeVar (Heap heap) (Var n tyrep) v' =
+    fromMaybe (error "writeVar: no such var") $ do
+    SomeValue v <- Map.lookup n heap
+    Refl        <- tyrep `eqTyRep` tyRepValue v
+    let heap' = Heap (Map.insert n (SomeValue v') heap)
+    return heap'
+
+-- | Extend the heap and allocs with a fresh variable.
+extendHeap :: (Heap, Allocs) -> Value t -> (Var t, Heap, Allocs)
+extendHeap (Heap heap, Heap allocs) v =
+    (var, Heap heap', Heap allocs')
+  where
+    var     = Var n' (tyRepValue v)
+    heap'   = Map.insert n' (SomeValue v) heap
+    allocs' = Map.insert n' (SomeValue v) allocs
+    n'     :: VarId
+    n'      = case Map.maxViewWithKey heap of
+                Nothing          -> VarId 0
+                Just ((n, _), _) -> succ n
+
+
+--
+-- Top level results
+--
+
+-- | The overall result of an STM transaction.
+--
+-- This is used for both the reference evaluator 'evalAtomically' and the
+-- conversion into the implementation STM via 'execAtomically'.
+--
+data TxResult =
+       TxComitted ImmValue
+     | TxBlocked
+     | TxAborted  ImmValue
+  deriving (Eq, Show)
+
+-- | An immutable snapshot of a 'Value' where the current values of the mutable
+-- variables are captured and included.
+--
+data ImmValue where
+
+    ImmValUnit ::             ImmValue
+    ImmValInt  :: Int      -> ImmValue
+    ImmValVar  :: ImmValue -> ImmValue
+    -- ^ This means it was the value within in a mutable variable, but the
+    -- identity of the variable is forgotten.
+  deriving (Eq, Show)
+
+-- | In the execution in real STM transactions are aborted by throwing an
+-- exception.
+--
+instance Exception ImmValue
+
+
+
+--
+-- Evaluation
+--
+
+evalExpr :: Env -> Expr t -> Value t
+evalExpr  env  (ExprName n) = lookupEnv env n
+evalExpr _env   ExprUnit    = ValUnit
+evalExpr _env  (ExprInt n)  = ValInt n
+
+-- | The normal form for a 'Term' after execution.
+--
+data NfTerm (t :: Type) where
+
+    NfReturn :: Value t -> NfTerm t
+    NfThrow  :: Value a -> NfTerm t
+    NfRetry  ::            NfTerm t
+
+deriving instance Show (NfTerm t)
+
+
+-- | The STM transition rules. They reduce a 'Term' to a normal-form 'NfTerm'.
+--
+evalTerm :: Env -> Heap -> Allocs -> Term t -> (NfTerm t, Heap, Allocs)
+evalTerm !env !heap !allocs term = case term of
```

Comment:
@dcoutts I would only add names of the rules.   This is very elegant translation of the operational semantics.

Reply:
Thanks :-)